pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1156 @@
+				r.nodeCIDRUpdateChannel <- workItem
+			}
+			r.lock.Unlock()
+		case <-stopChan:
+			return
+		}
+	}
+}
+
+func (r *multiCIDRRangeAllocator) cccWorker() {
+	defer raWaitGroup.Done()
+	for r.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (r *multiCIDRRangeAllocator) processNextWorkItem() bool {
+	cKey, quit := r.queue.Get()
+	if quit {
+		return false
+	}
+	defer r.queue.Done(cKey)
+
+	err := r.syncClusterCIDR(cKey.(string))
+	r.handleErr(err, cKey)
+
+	return true
+}
+
+func (r *multiCIDRRangeAllocator) handleErr(err error, key interface{}) {
+	if err == nil {
+		r.queue.Forget(key)
+		return
+	}
+
+	if r.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error syncing clusterCIDR %q, retries(%d). Error: %v", key, r.queue.NumRequeues(key), err)
+		r.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Error syncing ClusterCIDR %q, not retrying as retry budget (%d retries) has been exceeded: %v", key, maxRetries, err)
+	r.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+// createClusterCIDRHandler creates clusterCIDR handler.
+func createClusterCIDRHandler(f func(ccc *networkingv1alpha1.ClusterCIDR) error) func(obj interface{}) {
+	return func(originalObj interface{}) {
+		ccc := originalObj.(*networkingv1alpha1.ClusterCIDR)
+		if err := f(ccc); err != nil {
+			utilruntime.HandleError(fmt.Errorf("error while processing ClusterCIDR Add/Delete: %w", err))
+		}
+	}
+}
+
+// needToAddFinalizer checks if a finalizer should be added to the object.
+func needToAddFinalizer(obj metav1.Object, finalizer string) bool {
+	return obj.GetDeletionTimestamp() == nil && !slice.ContainsString(obj.GetFinalizers(),
+		finalizer, nil)
+}
+
+func (r *multiCIDRRangeAllocator) syncClusterCIDR(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing clusterCIDR request %q (%v)", key, time.Since(startTime))
+	}()
+
+	clusterCIDR, err := r.clusterCIDRLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		klog.V(3).Infof("clusterCIDR has been deleted: %v", key)
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// Check the DeletionTimestamp to determine if object is under deletion.
+	if !clusterCIDR.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(clusterCIDR)
+	}
+	return r.reconcileCreate(clusterCIDR)
+}
+
+func (r *multiCIDRRangeAllocator) insertNodeToProcessing(nodeName string) bool {
+	if r.nodesInProcessing.Has(nodeName) {
+		return false
+	}
+	r.nodesInProcessing.Insert(nodeName)
+	return true
+}
+
+func (r *multiCIDRRangeAllocator) removeNodeFromProcessing(nodeName string) {
+	klog.Infof("Removing Node: %s from processing", nodeName)
+	r.nodesInProcessing.Delete(nodeName)
+}
+
+// occupyCIDRs marks node.PodCIDRs[...] as used in allocator's tracked cidrSet.
+func (r *multiCIDRRangeAllocator) occupyCIDRs(node *v1.Node) error {
+
+	err := func(node *v1.Node) error {
```

Comment:
removeNodeFromProcessing must only be called after the Pod CIDRs have been occupied  for the node otherwise multiple workers might end up working on the same node